### ISSUE TITLE
Add a metrics module

### ISSUE THREAD

**Issue by rth** [FreeDiscovery/FreeDiscovery#20]
It might be useful to add a metrics module to FreeDiscovery so that given a truth file, the efficiency of clustering / duplicates detection can be checked solely with the REST API. Currently this requires to import metrics from scikit-learn. This would also give more flexibity for categorization scores,

This could consist of,
   1. Adding the necessary information retrieval specific metrics to [`freediscovery/metrics.py`](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/metrics.py) , the rest being imported form [`sklearn.metrics`](http://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) (at present there is nothing to be done in this step).
    2. Add a REST API endpoints  to `freediscovery/server/resources.py`,
       
            GET /api/v0/metrics/categorization
                  supported metrics:  'precision', 'recall', 'f1',  'roc_auc' 
            GET /api/v0/metrics/clustering
                  supported metrics: 'adjusted_rand', 'adjusted_mutual_info', 'v_measure', 'silhouette''
            GET /api/v0/metrics/duplicates-detection 
                  supported metrics: 'ratio_duplicates', 'f1_same_duplicates' 
                  (these ones are already defined in freediscovery)
         so that e.g. `GET /api/v0/metrics/categorization` with `data={'kind': ['f1', 'precision', 'recall'}` returned
        
             {'f1': <float>, 'precision': <float>, 'recall': <float>}
     3. Add unit-test to `test_api.py` and `test_metrics.py`

Over time, once #4 is implemented this would also allow to migrate from the [current limited scoring function](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/utils.py#L32) for categorization to this metrics module.

**Comment by DenisFLASH** [comment]
i can start working on this issue, if you don't mind

**Comment by DenisFLASH** [comment]
Just to be sure that i understood this issue correctly.

1. When running categorization example (LSI here), the post request to http://localhost:5001/api/v0/lsi/ad790e410cf7477da22c021b6c8b996d/test
gives the following 'res' object:
LSI res: {'precision': 0.04639175257731959, 'roc_auc': 0.8218507949449654, 'average_precision': 0.7515142921867958, 'recall': 0.75, 'f1': 0.08737864077669903}

It means that all categorization metrics we're interested in are being calculated using this endpoint (/lsi/<mid>/test). 
So my question is: the call to the endpoint you've proposed, GET /api/v0/metrics/categorization, 
should it just return the same metrics as above, with the only difference in that it can be filtered with data={'kind': ['f1', 'precision', 'recall'} to select only the metrics of interes, right?
In this case do we have to add <mid> to the endpoint , so that it becomes /api/v0/metrics/categorization/<mid> of the particular dataset ?

2. And for the other metrics we want to customize (not the standard scikit-learn metrics) ,  "metrics.py" file should be filled with implementations, right?

**Comment by rth** [comment]
Thanks, @DenisFLASH ! Well, maybe in this PR we could just wrap  a subset of `sklearn.metrics` listed above (and the ones that are already present in `freediscovery/metrics.py`) with an equivalent REST API. For instance, say [`sklearn.metrics.recall_score`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html#sklearn.metrics.recall_score)  takes as input two numeric arrays `y_true` and `y_pred`, and that's exactly what `GET /api/v0/metrics/categorization` should accept as input, with an additional parameter to compute several metrics with a single call. This would work well for dup. detection and clustering , as they currently have don't have a `/test` endpoint and require to manually import those metrics from python. For clustering / LSI the situation is indeed a bit more complicated, as we indeed have the `../test` endpoints, so what we could do is, for instance,
  1. In this PR just add those categorization metrics without changing the `../test` endpoints
  2. Address issue #4  (I'll look into it) with the numeric indexing of file names
  3. Afterwards, see how we can improve the `./test/` endpoints given the two points above.

So this PR could add additional endpoint at `/api/v0/metrics/...`, without affecting other endpoints, and we will see how to do the migration afterwards. 

 For metrics in `metrics.py` I think we can just wrap the ones that are implemented for the moment, there is no need to add other ones for now.

The end goal being that in future the only module aware of file names, or how to associate file names in ground truth files to the FreeDiscovery document id, should be the `/feature_extraction` module (with everything else based on numerical arrays or document ids). This should be a more standard way IMO, simplify the cross-OS compatibility, and make it easier to e.g. work with documents stored in a database (as opposed to a file system) if we ever need to do that in the future.  

What do you think?

**Comment by DenisFLASH** [comment]
Categorization metrics implemented and tested, pushed [here](https://github.com/DenisFLASH/FreeDiscovery/blob/metrics/freediscovery/server/resources.py)

Unit tests and CI passing OK. @rth , can you have a look and validate? if it's ok for you, i'll use the same logic for clustering and dup.detection metrics.

I just want to clarify the testing part. First i calculated all 4 categorization metrics for specific y_true and y_pred i used for a test (i used ipython to call sklearn metrics over those vectors). however, when i put those expected values in test code, one of assertions failed: f1 score, whose expected value was 0.75 , due to numerical reasons became 0.749999999999 when running tests. Which of these two options you suggest to deal with such cases:

- not using strict assertion, but introduce an "epsilon" with a needed precision (sort of "almostEquals")
- do not manually calculate expected values, but just call sklearn metrics in the test code, as the goal is not to test sklearn metrics itself, but rather the fact that our API calls exactly the corresponding sklearn metrics.  [I used this option for now](https://github.com/DenisFLASH/FreeDiscovery/blob/metrics/freediscovery/tests/test_server.py#L467), but looking forward to hearing your opinion.

**Comment by rth** [comment]
@DenisFLASH Yes, looks great! I think the first point would be simpler: for testing up to some number of decimals, we could use `assert_almost_equal` or `assert_allclose` from `numpy.testing`. 

Also would you mind wrapping the metrics calculations in,
     
```py
import warnings 
try:  # sklearn v0.17
      from sklearn.exceptions import UndefinedMetricWarning
except ImportError:  # v0.18
      from sklearn.metrics.base import UndefinedMetricWarning

with warnings.catch_warnings():
      warnings.simplefilter("ignore", category=UndefinedMetricWarning)
      # metrics calculations here
```
as in particular f1 score frequently prints warnings about "F-score is ill-defined and being set to 0.0 due to no predicted samples.". Thanks!

**Comment by DenisFLASH** [comment]
@rth , thanks for the answer. 
Pushed assert_almost_equals  and warnings wrapper [here](https://github.com/DenisFLASH/FreeDiscovery/tree/metrics)

I'll let you know when clustering and dup.detection metrics are ready.
By the way, concerning clustering. In the description of this issue you've listed 4 metrics  (supported metrics: 'adjusted_rand', 'adjusted_mutual_info', 'v_measure', 'silhouette'). The problem is that [silhouette_score ](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score) has another signature: instead of labels_true, labels_pred, it accepts as input (X, labels) where X is a feature array (pairwise distances between samples). So, 

- either it is possible to extract X array in order to pass it here, 
- or we can calculate silhouette_score only while doing clustering itself, and there will be no way to GET  
its value in the metrics API (in its current format).

What do you think?

**Comment by DenisFLASH** [comment]
update: 3 clustering metrics (adjusted_rand, adjusted_mutual_info, v_measure) [implemented and tested](https://github.com/DenisFLASH/FreeDiscovery/tree/metrics).

Concerning silhouette_score (see previous comment), looks like issue #26 has a bit in common with this distance array subject. So i think it will be a nice issue for me to assign, it could give a better understanding on how to deal with silhouette score.

Btw,  do we need **homogeneity** and **completeness**? we could add them in the list of possible cluster metrics, as their signature is (labels_true, labels_pred)

**Comment by rth** [comment]
@DenisFLASH Looks! I think we could just not add  the `silhouette_score`, we are not actually using it, besides it quite computationally expensive and requires a lot of RAM well last time I tried. 

I also think we can do without homogenity and completness scores as well (as `v_measure_score` is a harmonic mean of those and provides similar information). 

I'm not working on issue #26 at the moment so feel free to look into it. )

**Comment by DenisFLASH** [comment]
yes, `v_measure ` is a harmonic mean of those 2, like `f1_score` is a harmonic mean of `precision `and `recall`. I just thougth it could be useful in some cases, for the same reasons we do allow separetely get `precision ` and `recall`, even if F1 is also calculated.

Ok for #26

**Comment by DenisFLASH** [comment]
Finished with dup.detection metrics, pushed, created PR ( #36 )
@rth,  i'll close this issue after you accept the PR

**Comment by DenisFLASH** [comment]
@rth just added documentation and [pushed it](https://github.com/DenisFLASH/FreeDiscovery/tree/metrics-docs/doc/rest_api)  How can i check new documentation locally before creating a PR?

**Comment by rth** [comment]
Great! It's documented here https://freediscovery.github.io/doc/dev/contributing.html#building-documentation You also have to start the server at localhost with `bash ./conda_run.sh` in a separate terminal, so that examples can be built for the documentation. The documentation build might be Linux/Mac OS only as well (not sure about this). Feel free to improve that section too if it's not sufficiently documented ) 

Also Circle CI is responsible for building the documentation but you have to make a PR for it be executed. The updating of online documentation is also done by Circle CI but only for commits on the FreeDiscovery master branch (i.e. once the PR is merged).

**Comment by DenisFLASH** [comment]
Still can't build the doc, as `make html` gives me import errors of `sphinx_rtd_theme`, then `recommonmark`.  I added sphinx==1.3.3 in ``equirements_pip_win.txt`, now i get `could not import extension matplotlib.sphinxext.plot_directive`.
Looks like for Windows this building docs procedure will really be painful. Maybe i can make a PR of[ this branch](https://github.com/DenisFLASH/FreeDiscovery/tree/metrics-docs) which implements documentation for already existing Metrics module, and then you'll build it on Linux to verify and accept PR? Just let me know if it is ok, and i'll create PR.  Otherwise, i have to continue growing requirements file

**Comment by rth** [comment]
for building docs you need to install `pip -r build_tools/requirements_extra_pip.txt` (cf that documentation page), that should give you all the dependencies...

**Comment by DenisFLASH** [comment]
completely forgot abour `requirements_extra_pip`, now it's okay

i had a windows-related problem because of my Anaconda and FreeDiscovery folders were on two different drives (C and D) and it caused HTML generation to stop with an error in "ntpath.py" file. Nowi copied this folder to the same C drive, it works well.

Pull request #40 is created

**Comment by rth** [comment]
Fixed in PR #40.
